File: client/src/main/java/au/com/codeka/warworlds/client/store/ProtobufStore.java
package au.com.codeka.warworlds.client.store;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import com.squareup.wire.Message;

import java.util.Map;

import javax.annotation.Nonnull;

/** Class for storing protos in a sqlite key-value store, keyed by a long id. */
public class ProtobufStore<M extends Message<?, ?>> extends BaseStore<Long, M> {
  private final String name;
  private final ProtobufSerializer<M> serializer;
  private final SQLiteOpenHelper helper;

  public ProtobufStore(String name, Class<M> protoClass, SQLiteOpenHelper helper) {
    super(name, helper);
    this.name = name;
    this.helper = helper;
    this.serializer = new ProtobufSerializer<>(protoClass);
  }

  @Override
  public void onCreate(SQLiteDatabase db) {
    db.execSQL(
        "CREATE TABLE " + name + " ("
            + "  key INTEGER PRIMARY KEY,"
            + "  value BLOB)");
  }


  /**
   * Gets the value with the given ID, or {@code null} if it's not found.
   */
  public M get(long key) {
    SQLiteDatabase db = helper.getReadableDatabase();
    try (Cursor cursor = db.query(false, name, new String[]{ "value" },
        "key = ?", new String[]{ Long.toString(key) },
        null, null, null, null)) {
      if (cursor.moveToFirst()) {
        return serializer.deserialize(cursor.getBlob(0));
      }
    }

    return null;
  }

  /**
   * Puts the given value to the data store.
   */
  public void put(long id, @Nonnull M value) {
    SQLiteDatabase db = helper.getWritableDatabase();
    ContentValues values = new ContentValues();
    values.put("key", id);
    values.put("value", serializer.serialize(value));
    db.insertWithOnConflict(name, null, values, SQLiteDatabase.CONFLICT_REPLACE);
  }

  /** Puts all of the given values into the data store in a single transaction. */
  public void putAll(Map<Long, M> values) {
    SQLiteDatabase db = helper.getWritableDatabase();
    db.beginTransaction();
    try {
      ContentValues contentValues = new ContentValues();
      for (Map.Entry<Long, M> kvp : values.entrySet()) {
        contentValues.put("key", kvp.getKey());
        contentValues.put("value", serializer.serialize(kvp.getValue()));
        db.insertWithOnConflict(name, null, contentValues, SQLiteDatabase.CONFLICT_REPLACE);
      }
      db.setTransactionSuccessful();
    } finally {
      db.endTransaction();
    }
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/store/ProtobufStore.kt
package au.com.codeka.warworlds.client.store

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.squareup.wire.Message

/** Class for storing protos in a sqlite key-value store, keyed by a long id.  */
class ProtobufStore<M : Message<*, *>?>(
    private val name: String, protoClass: Class<M>, private val helper: SQLiteOpenHelper)
  : BaseStore(name, helper) {
  private val serializer: ProtobufSerializer<M>
  override fun onCreate(db: SQLiteDatabase?) {
    db!!.execSQL(
        "CREATE TABLE " + name + " ("
            + "  key INTEGER PRIMARY KEY,"
            + "  value BLOB)")
  }

  /**
   * Gets the value with the given ID, or `null` if it's not found.
   */
  operator fun get(key: Long): M? {
    val db = helper.readableDatabase
    db.query(false, name, arrayOf("value"),
        "key = ?", arrayOf(java.lang.Long.toString(key)),
        null, null, null, null).use { cursor ->
      if (cursor.moveToFirst()) {
        return serializer.deserialize(cursor.getBlob(0))
      }
    }
    return null
  }

  /**
   * Puts the given value to the data store.
   */
  fun put(id: Long, value: M) {
    val db = helper.writableDatabase
    val values = ContentValues()
    values.put("key", id)
    values.put("value", serializer.serialize(value))
    db.insertWithOnConflict(name, null, values, SQLiteDatabase.CONFLICT_REPLACE)
  }

  /** Puts all of the given values into the data store in a single transaction.  */
  fun putAll(values: Map<Long?, M>) {
    val db = helper.writableDatabase
    db.beginTransaction()
    try {
      val contentValues = ContentValues()
      for ((key, value) in values) {
        contentValues.put("key", key)
        contentValues.put("value", serializer.serialize(value))
        db.insertWithOnConflict(name, null, contentValues, SQLiteDatabase.CONFLICT_REPLACE)
      }
      db.setTransactionSuccessful()
    } finally {
      db.endTransaction()
    }
  }

  init {
    serializer = ProtobufSerializer(protoClass)
  }
}